flowmachine/flowmachine/features/subscriber/entropy.py

```diff
@@ -0,0 +1,391 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# -*- coding: utf-8 -*-
+"""
+Calculates various entropy metrics for subscribers with a specified time
+period.
+"""
+
+from abc import ABCMeta
+from .metaclasses import SubscriberFeature
+from .contact_balance import ContactBalance
+from ..utilities.sets import EventsTablesUnion
+from ..utilities.subscriber_locations import subscriber_locations
+from ...utils.utils import get_columns_for_level
+from ...core import Table
+
+
+class BaseEntropy(SubscriberFeature, metaclass=ABCMeta):
+    """ Base query for calculating entropy of subscriber features. """
+
+    def _make_query(self):
+
+        return f"""
+        SELECT
+            subscriber,
+            -1 * SUM( relative_freq * LN( relative_freq ) ) AS entropy
+        FROM ({self._relative_freq_query}) u
+        GROUP BY subscriber
+        """
+
+    @property
```

Comment:
Suggest to add an `@abstractproperty` here, which forces any subclass to implement it (and which is detected during the parsing stage, otherwise you'll actually have to run the query to hit the NotImplementedError).
```suggestion
    @property
    @abstractmethod
```